### ISSUE TITLE
PhaseEncoding direction converted to a complex number

### ISSUE THREAD

**Issue by jan-petr** [ExploreASL/ExploreASL#378]
PhaseEncoding direction is still saved and loaded as a complex number. In certain cases. See the example error below:

This was fixed previously for some cases, but not all in #306 

Converting from BIDS to Legacy: /pet/projekte/asl/data/BIDS/TmpConversion/GE_PCASL_2DEPI_volunte100%
Warning: Complex numbers not supported 
> In spm_jsonwrite>jsonwrite_numeric (line 198)
  In spm_jsonwrite>jsonwrite_var (line 91)
  In spm_jsonwrite>jsonwrite_struct (line 156)
  In spm_jsonwrite>jsonwrite_var (line 81)
  In spm_jsonwrite (line 61)
  In xASL_bids_parseM0 (line 51)
  In xASL_bids_BIDS2Legacy (line 254)
  In xASL_test_BIDSFlavorsFull (line 149)
  In BIDS_fullPipelineTest (line 47)

### Comment
Reverts commits from issue #306. The complex number issue is fixed at a lower-level when reading JSON, not at higher level when interpreting the fields. Also - removed parts of code that were commented. See also #62 for future ideas to fix these issues.

### Release text
Fix bug in reading PhaseEncodingDirection in JSON as a complex number instead of a char.

**Comment by MichaelStritt** [comment]
Here 00b248a we just fixed this in in `xASL_adm_LoadParms` by converting the PhaseEncodingAxis from numeric `i` or `-i` to a char `'i'` or `'-i'`. As far as I can see, here the error is that we write a numeric complex number using jsonwrite. This means here we probably should make sure that we give a char (`'i'` or `'-i'`) for the PhaseEncodingAxis to jsonwrite, correct?

**Comment by MichaelStritt** [comment]
@jan-petr: I just ran a few tests. So the **JSON** file is read in `xASL_bids_parseM0(PathNifti)`.

So in line 111 of `[x] = xASL_import_json(DataParFile)` we convert **strings/chars** to **numbers**:

```
x = ConvertNumericalFields(x);
```

Here `x.PhaseEncodingDirection` is converted from `"j"` to `"0+1i"`. Does `x.PhaseEncodingDirection` have to be a number in ExploreASL? Do we have to convert it? (Sorry, I couldn't find anything about it in the DataParFile template). Writing this complex number back to the **JSON** leads to the shown warning.

**Comment by MichaelStritt** [comment]
I think I know the fix, we remove the conversion in `xASL_adm_LoadParms.m` and just check if the converted number would be imaginary in `x = ConvertNumericalFields(x);`, if this is the case we don't do the conversion, right?

**Comment by MichaelStritt** [comment]
Sorry Jan, I think I need your feedback here.

### 1. Question
Do we need the conversion from str2num for anything besides the vectors? Do we even need it for this?

### 2. Question
...or can we just get rid of `x = ConvertNumericalFields(x);` in `[x] = xASL_import_json(DataParFile)` completely?

### 3. Question
Is there a maximum number of fields or why did we add this part?
```
if length(FieldsAre{iField})>63
    warning('Invalid field name, removing from struct');
else
...
```

I removed the "fix" of #306 in e405168 and added a check for complex numbers, but if we need to be able to read complex numbers, this is of course not correct.

**Comment by jan-petr** [comment]
> I think I know the fix, we remove the conversion in `xASL_adm_LoadParms.m` and just check if the converted number would be imaginary in `x = ConvertNumericalFields(x);`, if this is the case we don't do the conversion, right?

Yes. I totally agree. This should be solved in the I/O functions and not in the higher-level functions. Thanks for finding out all this. I will take over from now to check if all is done correctly.

The values can be 'i','-i','j','-j','k','-k'. Both the sign and the character matters - it displays both the dimension and the direction.

**Comment by jan-petr** [comment]
I will implement those, but FYI:

> ### 1. Question
> Do we need the conversion from str2num for anything besides the vectors? Do we even need it for this?

Yes. Unfortunately, some character arrays can be entered as numbers - some people/software were doing this, so we better keep it. Complex numbers are the only issue, so we can make this exception to be sure that it all works. And we can hopefully phase this out in the future when we are sure that numbers in vectors do not occur anymore.

> ### 2. Question
> ...or can we just get rid of `x = ConvertNumericalFields(x);` in `[x] = xASL_import_json(DataParFile)` completely?

As said above, not yet.

> ### 3. Question
> Is there a maximum number of fields or why did we add this part?
> 
> ```
> if length(FieldsAre{iField})>63
>     warning('Invalid field name, removing from struct');
> else
> ...
> ```

No idea why this is there. Will check. Thanks, that can be a potential source of errors.

> I removed the "fix" of #306 in [e405168](https://github.com/ExploreASL/ExploreASL/commit/e405168c9c06e63f7dc54d988611869ee30a6e43) and added a check for complex numbers, but if we need to be able to read complex numbers, this is of course not correct.

Perfect, thanks. No, I do not think we need complex numbers. Not in the JSON. If this ever occurs, we will extend this behavior to specific fields only.

**Comment by jan-petr** [comment]
> > ### 3. Question
> > Is there a maximum number of fields or why did we add this part?
> > ```
> > if length(FieldsAre{iField})>63
> >     warning('Invalid field name, removing from struct');
> > else
> > ...
> > ```

Sorry - this is of course not the number of fields, but length of the field name. Still - there are no hard limits on that. Definitively not 63.
